plugin/rewrite/reverter.go

```diff
@@ -21,23 +22,69 @@ type ResponseRule struct {
 // dig will complain with ';; Question section mismatch: got example.org/HINFO/IN'
 type ResponseReverter struct {
 	dns.ResponseWriter
-	originalQuestion dns.Question
-	ResponseRewrite  bool
-	ResponseRules    []ResponseRule
+	request           *dns.Msg
+	originalQuestion  dns.Question
+	originalOPTHeader dns.RR_Header
+	originalHasOPT    bool
+	ResponseRewrite   bool
+	ResponseRules     []ResponseRule
+}
+
+// align the OPT inside the msg the information provided - hasOPT and Header
+// the sub array of EDNS0 is not changed (unless OPT is removed)
+func replaceOPTHeader(m *dns.Msg, hasOpt bool, hdr dns.RR_Header) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if !hasOpt {
+				copy(m.Extra[i:], m.Extra[i+1:])
+				m.Extra[len(m.Extra)-1] = nil
+				m.Extra = m.Extra[:len(m.Extra)-1]
+				return
+			}
+			m.Extra[i].(*dns.OPT).Hdr = hdr
+			return
+		}
+	}
+	if hasOpt {
+		opt := dns.OPT{hdr, make([]dns.EDNS0, 0)}
+		m.Extra = append(m.Extra, &opt)
+	}
 }
 
 // NewResponseReverter returns a pointer to a new ResponseReverter.
 func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg) *ResponseReverter {
+	hdr := dns.RR_Header{}
+	o := r.IsEdns0()
+	if o != nil {
+		hdr = *o.Header()
+	}
 	return &ResponseReverter{
-		ResponseWriter:   w,
-		originalQuestion: r.Question[0],
+		ResponseWriter:    w,
+		request:           r,
+		originalQuestion:  r.Question[0],
+		originalHasOPT:    o != nil,
+		originalOPTHeader: hdr,
 	}
 }
 
 // WriteMsg records the status code and calls the underlying ResponseWriter's WriteMsg method.
 func (r *ResponseReverter) WriteMsg(res *dns.Msg) error {
-	res.Question[0] = r.originalQuestion
+
+	// set back the client request to its original OPT header
+	replaceOPTHeader(r.request, r.originalHasOPT, r.originalOPTHeader)
+	if !r.originalHasOPT {
+		// ensure remove OPT out of the response
+		// if initial query was not EDNS0 - compliance with https://tools.ietf.org/html/rfc6891#page-10
+		replaceOPTHeader(res, false, r.originalOPTHeader)
+	}
+
+	// Revert part happen ONLY if flag is set
 	if r.ResponseRewrite {
+
+		// always revert the question
+		res.Question[0] = r.originalQuestion
```

Comment:
I believe this should be unconditional. The question name in response must be the same as in request

Reply:
You are right.
But here it is messy. 
- "noRevert" is only a UT case. and is used only to test if rewrite act as expected
- other case (with a response) needs to revert the OPT and the question
- in those other cases, some option requires a  "responseRewrite" (the rules that are Active)

I will change back to specific case for "noRevert" and let the use of Reverter only for real response use case.